sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,530 @@
+                else ret = cipher.doFinal(tmpBuffer, dst);
+            } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException ex) {
+                throw new IOException("failed to decrypt blck.", ex);
+            }
+            return ret;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected void readInternal(ByteBuffer b) throws IOException {
+            long pos = getFilePointer() + off;
+
+            if (pos + b.remaining() > end) {
+                throw new EOFException(
+                    Thread.currentThread().getId()
+                        + " read past EOF: "
+                        + this
+                        + " isClone? "
+                        + isClone
+                        + " off: "
+                        + off
+                        + " pos: "
+                        + pos
+                        + " end: "
+                        + end
+                );
+            }
+
+            try {
+                int readLength = b.remaining();
+                while (readLength > 0) {
+                    final int toRead = Math.min(CHUNK_SIZE, readLength);
+                    b.limit(b.position() + toRead);
+                    assert b.remaining() == toRead;
+                    final int i = read(b, pos);
+                    if (i < 0) {
+                        throw new EOFException("read past EOF: " + this + " buffer: " + b + " chunkLen: " + toRead + " end: " + end);
+                    }
+                    assert i > 0 : "FileChannel.read with non zero-length bb.remaining() must always read at least "
+                        + "one byte (FileChannel is in blocking mode, see spec of ReadableByteChannel)";
+                    pos += i;
+                    readLength -= i;
+                }
+                assert readLength == 0;
+            } catch (IOException ioe) {
+                throw new IOException(ioe.getMessage() + ": " + this, ioe);
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected void seekInternal(long pos) throws IOException {
+            if (pos > length()) {
+                throw new EOFException(
+                    Thread.currentThread().getId() + " read past EOF: pos=" + pos + " vs length=" + length() + ": " + this
+                );
+            }
+            CipherFactory.initCipher(cipher, directory, Optional.empty(), Cipher.DECRYPT_MODE, pos + off);
+        }
+    }
+
+    /**
+     * An IndexOutput implementation that encrypts data before writing
+     *
+     * @opensearch.internal
+     */
+    final class CryptoIndexOutput extends OutputStreamIndexOutput {
+        /**
+         * The maximum chunk size is 8192 bytes, because file channel mallocs a native buffer outside of
+         * stack if the write buffer size is larger.
+         */
+        static final int CHUNK_SIZE = 8192;
+
```

Comment:
It looks like this is half of the read size above. Should these be the same?

Reply:
They don't have to be the same size for the encryption/decryption to be done correctly. These sizes are the defaults chosen by lucene.